tools/node-modules

```diff
@@ -56,35 +56,96 @@ EOF
     clone_from_cache "${sha}"
 }
 
-cmd_rebuild() {
-    for file in npm-shrinkwrap.json package-lock.json yarn.lock; do
-        if [ -e "${file}" ]; then
-            echo "Please delete ${file} from $(pwd) and start again"
-            exit 1
-        fi
-    done
+cmd_npm_install() {
+    if [ -f /run/.toolboxenv ]; then
+        exec flatpak-spawn --host -- "$0" npm_install "$@"
+        exit 1
+    fi
+
+    CACHE="cockpit-project-npm-cache-volume"
+    IMAGE="docker.io/library/node:alpine"
+
+    # make sure the node user can write to the cache volume
+    podman run \
+        --rm \
+        --volume "${CACHE}":/home/node/.npm:U \
+        "${IMAGE}" chown -R node:node /home/node >&2
+
+    # do the actual work
+    podman run \
+        --log-driver='none' \
+        --rm \
+        --user node \
+        --workdir /home/node \
+        --interactive \
+        --attach stdin \
+        --attach stdout \
+        --attach stderr \
+        --volume "${CACHE}":/home/node/.npm \
+        "${IMAGE}" /bin/sh -c '
+            set -eux
+            tee package.json >/dev/null
+            npm install --ignore-scripts >&2
+            cp package.json node_modules/.package.json
+            tar c node_modules
+        '
+}
+
+cat_file() {
+    case "$1" in
+        *:*)
+            git cat-file blob "$1";;
+        *)
+            cat "$1";;
+    esac
+}
 
+tar_to_git() {
+    tar --exclude '.git*' --extract "$1"
+    message INDEX "node_modules"
+    git -C "$1" init --quiet --initial-branch=main
+    git -C "$1" add --all
+}
+
+cmd_install() {
     cmd_remove
 
-    npm install
+    cmd_npm_install < package.json | tar_to_git node_modules
 
-    # try to keep this vaguely in sync with what .github/workflows/node_modules-* do
-    cp package.json node_modules/.package.json
-    cp package-lock.json node_modules/.package-lock.json
-    echo /.cache > node_modules/.gitignore
+    # consistency check
+    cmp package.json node_modules/.package.json
+    < node_modules/.package-lock.json
```

Comment:
Can this be one line, please?

Reply:
This is one line, although apparently entirely unobvious what I'm trying to do here.  Those are two entirely separate commands.

Again, this is about the side effects, but for `set -e` in this case.  I want the script to fail if the `.package-lock.json` file is missing, and I want it to fail with a helpful error message.  Doing the usual `test -f ...` doesn't issue an error if it doesn't exist, unfortunately...

Again, open to better suggestions.  I guess I was trying to avoid open-coding the entire `if test ...; then echo ... 2>&1; exit 1; fi` ordeal...

Reply:
Oh right, I mis-looked. It felt like calling `cmp` against a file and stdin, but that's of course not what happens.

How about `ls -l node_modules/.package-lock.json`? For anyone looking it might be a nice spot-check (file size and such), and for non-existing files it gives an appropriate error message.

Reply:
Thanks for the suggestion.  `>&2` of course.